### ISSUE TITLE
Hi, I have a question about preprocessing!

### ISSUE THREAD

**Issue by Yoontae6719** [s-marton/GRANDE#2]
Hi, I read your recent research with great interest. Thank you, but I have a question about the process of Quantile transform in your code. It seems to me that in this process, the X_data is before the train-test split, so the information about the test set is used together. Looking forward to your answer.

```
def normalize_data(X_data, normalizer_list=None, technique='min-max', low=-1, high=1, quantile_noise=1e-3, random_state=42, exclude_columns=[]):
    if normalizer_list is None:
        normalizer_list = []
        if isinstance(X_data, pd.DataFrame):
            if technique != 'quantile':
            
                for column_name in X_data:
                    if column_name not in exclude_columns:
                        if technique == 'min-max':
                            scaler = MinMaxScaler(feature_range=(low, high))
                        elif technique == 'mean':
                            scaler = StandardScaler()

                        scaler.fit(X_data[column_name].values.reshape(-1, 1))

                        X_data[column_name] = scaler.transform(X_data[column_name].values.reshape(-1, 1)).ravel()
                        normalizer_list.append(scaler)
                
            else:
                
                Z_data = X_data.drop(exclude_columns, axis=1)
                relevant_columns = Z_data.columns
                
                if len(relevant_columns) > 0:

                    quantile_train = np.copy(Z_data.values).astype(np.float64)
                    if quantile_noise > 0:
                        np.random.seed(random_state)
                        stds = np.std(quantile_train, axis=0, keepdims=True)
                        noise_std = quantile_noise / np.maximum(stds, quantile_noise)
                        quantile_train += noise_std * np.random.randn(*quantile_train.shape)                        

                    scaler = QuantileTransformer(output_distribution='normal', random_state=random_state)
                    scaler.fit(quantile_train)   
                    Z_data = pd.DataFrame(scaler.transform(Z_data.values), columns=Z_data.columns, index=Z_data.index)

                    X_data = pd.concat([X_data[exclude_columns], pd.DataFrame(Z_data, index=X_data.index, columns=relevant_columns)], axis=1)

                    normalizer_list.append(scaler) 
```

**Comment by s-marton** [comment]
Hi @Yoontae6719,
first of all thank you very much for the feedback! 

I looked into your issue. While the notation might be a bit confusing, the normalization is applied **after** the train-test split and therefore only the training data is used for preprocessing. An exemplary usage from ```utilities_GRANDE.py``` is shown below. When passing the training data, the normalization parameters are calculated and returned as ```normalizer_list```. In the following calls for the validation and test data, these parameters are passed to the ```normalize_data``` function as an additional argument and used for preprocessing.

```
X_train, normalizer_list = normalize_data(X_train, technique=config['preprocessing']['normalization_technique'], quantile_noise=config['preprocessing']['quantile_noise'], exclude_columns=exclude_columns)
X_valid, _ = normalize_data(X_valid, normalizer_list=normalizer_list, technique=config['preprocessing']['normalization_technique'], quantile_noise=config['preprocessing']['quantile_noise'], exclude_columns=exclude_columns)
X_test, _ = normalize_data(X_test, normalizer_list=normalizer_list, technique=config['preprocessing']['normalization_technique'], quantile_noise=config['preprocessing']['quantile_noise'], exclude_columns=exclude_columns)
```

In the updated version used in the pip package and the main part of the repo, this functionality is included more clearly in the model using the function ```perform_preprocessing``` which is called before training the model and  ```apply_preprocessing``` to preprocess unseen data for prediction.
